diff -urN elspy-0.1.1.org/elspy.c elspy-0.1.1/elspy.c
--- elspy-0.1.1.org/elspy.c	2005-09-01 18:14:46.800078888 +0200
+++ elspy-0.1.1/elspy.c	2005-09-01 18:14:12.000000000 +0200
@@ -578,6 +578,8 @@
    PyObject * lscan_return;
    int        result;
 
+   putenv("PYTHONPATH=/etc/mail");
+	   
    if (! py_initialized) {
       log_debug(2, "initializing Python interpreter and _elspy module");
       Py_Initialize();
